### ISSUE TITLE
Clarify MRSP 5.3 Requirement to Disclose Self-signed Certificates

### ISSUE THREAD

**Issue by wthayer** [mozilla/pkipolicy#186]
Policy section 5.3 is titled "Intermediate Certificates" which implies that it does not apply to roots. The first paragraph requires disclosure of such certificates.

It is common practice to cross-sign new roots with older, more ubiquitous ones. This creates a chain like:
leaf-->new intermediate-->new cross cert (key A + subject A, issuer B)-->old root (key B, subject B)

What is not obvious, and has tripped up a number of CAs, is the fact that the cross cert also enables the following path:

leaf-->new intermediate-->new root (key A + subject A, issuer A)-->new cross cert (key A + subject A, issuer B)-->old root (key B, subject B)

In other words, the new root meets the section 5.3 definition of an intermediate that must be disclosed.

I propose that this requirement be clarified by adding a sentence like: "This includes self-signed certificates that share a Subject and public key with another certificate subject to this requirement."

Background: https://bugzilla.mozilla.org/show_bug.cgi?id=1563575

**Comment by sleevi** [comment]
Language proposal looks good to me.

I had some initial hesitation with it, which I think was due to thinking about certificates based on "Issuer" rather than "Subject", but the more I read it and thought about it, the more I realized how perfectly it captures the requirements. I also am convinced it's better to refer to Subject, rather than Issuer, to avoid any confusion.

**Comment by dzacharo** [comment]
If I recall, only CCADB administrators can add self-signed RootCA Certificates to CCADB. Perhaps the process of adding a self-signed RootCA Certificate by the CA, should be described in https://ccadb.org/policy

**Comment by dzacharo** [comment]
@wthayer, I believe the second path you described is not going to take the "new root" into consideration.

**Comment by sleevi** [comment]
They can upload and disclose just fine. They cannot create a new root case
or treat it as a root record.

Can you explain why you do not believe it takes something in consideration?

**Comment by dzacharo** [comment]
I read the CCADB policy. A self-signed root CA Certificate is described in section 3, Root Certificates. It cannot be considered an Intermediate CA Certificate. A cross-certificate IS an intermediate CA Certificate so if the issuing RootCA is in CCADB and part of the program, the cross-certificate must be disclosed according to the existing policy. I see no ambiguity there. I'd be curious to know how @kwilson treated similar cases in the past, because adding a self-signed RootCA Certificate as an Intermediate CA Certificate doesn't seem right to me but more like a way to bypass restrictions to add such a certificate as a RootCA certificate.

**Comment by sleevi** [comment]
We have handled this case multiple times.

It is unclear if you are objecting to the proposed language (as a means to
accomplish a goal), or if you are objecting to the goal itself. The
objections seem consistent with objecting to the goal, so it might help to
clarify.

RFC 5280 makes it clear the ontology if certificates. It sounds like you’re
unhappy with that definition. The point is that such certificates
transitively chain to Mozilla-included roots if they are self-signed (which
by definition also means self-issued) and there exists a cross-certificate
bearing the same name and subject. By definition, that also makes them
intermediates in a certificate path.

**Comment by dzacharo** [comment]
In Wayne's second example, is the new root CA certificate adding something to the chain building?

**Comment by sleevi** [comment]
I’m not sure what you mean by added to the chain building. It is a valid
path that can be constructed.

The CCADB policy speaks in terms of trust anchors, which is a request the
CA specifically makes.

Regardless of them making the request or not, certificates which directly
or transitively chain to a Mozilla-included root should be disclosed.

Such certificates directly or transitively chain to Mozilla-included roots,
are capable of being disclosed in CCADB, we have repeatedly required
disclosure in CCADB, and tools that examine certificates for disclosures
(such as crt.sh) detect and report such certificates. They are self-signed,
thus self-issued, but such certificates are valid paths. They may be
suboptimal for performance, but that does not mean they do not chain to
Mozilla-included roots.

**Comment by dzacharo** [comment]
Perhaps I am confused by my own ignorance so please bare with me while trying to write an example.

Case 1:
Leaf (subjectDN=example.com, issuerDN=newIssuingCA) - - > New Issuing CA (subjectDN=newIssuingCA, issuerDN=new Root) - - > Cross Certificate CA (subjectDN=new Root, issuerDN=old Root) - - > old Root (subjectDN=issuerDN=old Root= Trust Anchor).

Isn't this the effective chain path for a client that only contains the old Root as a Trust Anchor?

Case 2:
Use Case 1 but also include the new self-signed root certificate in the chain. 

If the client doesn't have the new Root as a Trust Anchor, do you see adding the new self-signed RootCA (subjectDN=issuerDN=new Root) in the Certificate chain path to effectively be used in the trust building?

I guess this is my main question. Adding such a certificate (the new self-signed root) in the path when the client only has the old root as a trust anchor seems unnecessary and unusable.

**Comment by sleevi** [comment]
Yes, the path Wayne described:
leaf-->new intermediate-->new root (key A + subject A, issuer A)-->new cross cert (key A + subject A, issuer B)-->old root (key B, subject B)

Is a perfectly valid certificate path. It's not the ideal path, but it's a valid path, and thus you can say "new root" transitively chains to "old root" and that "new root" is being used as an intermediate.

**Comment by sleevi** [comment]
I think the mistake or confusion may come from thinking about disclosure in the sense of "The single path a client builds", when instead, disclosure is supposed to be thought of "The sum total of all possible paths through the graph". It's important to remember that the certificate graph is a directed, cyclic graph - many paths may exist simultaneously through the graph to a trust anchor.

**Comment by techliaison** [comment]
should it make a difference on whether or not the self-signed new root cert is publicly discoverable for an RP?  If the new root is not yet being distributed by any trust store and is not discoverable in an AIA to be used, just because it has been created, why should it need to be disclosed?  The new key has been disclosed by way of the cross-certificate which is what would be used if it issued any certificates.

**Comment by sleevi** [comment]
Mozilla Policy already clarifies this: disclosure is tied to creation, not
discoverability. So yes, it would be required to be disclosed after
creation.

We want to have the graph that exists disclosed, including those that have
been misissued or might not yet be in use, as they all represent risks to
the ecosystem.

Is there a reason a CA should be able to cover up a certificate that it has
created, temporarily or indefinitely? How would that help benefit users?

**Comment by dzacharo** [comment]
Just to clarify, I never objected to the requirement to disclose the new self-signed Root CA Certificate. I object to disclosing it as an Intermediate CA Certificate and I would prefer a process in ccadb where each CA can disclose a new self-signed RootCA Certificate with the assistance (or not) of Root program managers.

**Comment by techliaison** [comment]
I actually do not understand how the self-signed cert can even be used as an intermediate so I think I agree with Dimitris.

**Comment by sleevi** [comment]
@dzacharo Thanks for clarifying your position. Could you explain it as well? I thought your concerns were that you didn't think it was supported - but it is, and it's perfectly fine, and perfectly desirable. There are plenty of "CAs" (Subject + SPKI) that have multiple disclosures attached to them, sometimes by different CAs. If it doesn't benefit the CCADB maintainers, it's not clear to me how it benefits CAs to *not* disclose something.

@techliaison I'm not sure what you mean? It's a perfectly valid RFC 5280 path. RFC 6.1.3's algorithm goes out of the way to permit such certificates

**Comment by eabalea** [comment]
It's indeed a perfectly valid certificate path, according to RFC5280 and X.509.
But often hidden by software because of the confusion between "self-signed" and "trust anchor".

The self-signed certificate "new root (key A + subject A, issuer A)" is also a self-issued certificate, issued by "new cross cert (key A + subject A, issuer B)".

As long as the CA manager didn't mess in the different possible constraints, both paths are valid.

**Comment by techliaison** [comment]
I guess I am more used to path validation also checking the AKI vs just Issuer DN - but relooking at 5280, I see that this is not mandatory.

I do not see that 6.1.3 goes out of its way to discuss self-signed - just self-issued.  5280 makes a distinction between self-signed & self-issued.  3.2 states: Self-signed certificates are used to convey a public key for use to begin certification paths.

However, I still feel that certs are not issued to other certificates so if there is no way for the RP application to discover the next cert in the path even though it can be forced into a path doesn't make it an actual intermediate certificate and disclosing it as an intermediate may cause more confusion rather than help.

**Comment by sleevi** [comment]
Self-signed certificates are a subset of self-issued certificates.

I disagree with the framing on "issued". Perhaps we could resolve the ambiguity by stating that all certificates which have an "Issuer" field equivalent to a "Subject" field in a certificate subject to Mozilla policy, and which are signed by the private key associated to the issuer's public key, must be disclosed.

This functionally avoids mentioning self-signed, so I think is perhaps more confusing, but perhaps makes it clearer _why_ such certificates are required to be disclosed: Because they've been issued by a certificate subject to Mozilla requirements.

**Comment by dzacharo** [comment]
> @dzacharo Thanks for clarifying your position. Could you explain it as well? I thought your concerns were that you didn't think it was supported - but it is, and it's perfectly fine, and perfectly desirable. There are plenty of "CAs" (Subject + SPKI) that have multiple disclosures attached to them, sometimes by different CAs. If it doesn't benefit the CCADB maintainers, it's not clear to me how it benefits CAs to _not_ disclose something.

I am a bit confused with the multiple negatives in your previous comment. I have to dig back to see Kathleen's instructions but if my memory serves me well, she wanted CAs that participate in the Mozilla program to disclose all existing paths that chain to Roots in the program. If all members of the program performed this action, then all cross-certificates with paths that "transitively" chain to self-signed Roots served as Trust Anchors by Mozilla products would be disclosed. For cases where a CA created a new Root CA, it was not part of her instructions but CAs that wanted this included in CCADB would just notify her and she would add this information to the database without being marked as "Trusted" (yet).

I support the disclosure of self-signed Roots but with some clear instructions to notify a CCADB manager so it can be added to the database as a candidate RootCA. Perhaps my assumption is incorrect but this is how I understand the distinction between a self-signed RootCA and an Intermediate CA Certificate.

**Comment by sleevi** [comment]
@dzacharo Thanks for clarifying the concern.

There are two, wholly separate, wholly independent things:
- The disclosure of all certificate paths that chain to existing Mozilla-included roots
- A request for the inclusion of a certificate as a trust anchor

The two are entirely independent actions. A CA MUST disclose all certificate paths if they have a Mozilla-included root. They use CCADB to do so, and upload all the paths, including any self-signed, self-issued certificate that transitively chain to an existing root.

A CA MAY wish to have _any_ certificate treated as a trust anchor. Mozilla does not require that they be self-signed, self-issued "root" certificates (although that is preferable). For that, you need a CCADB member to open a Case to have that certificate treated as a trust anchor/"root".

As @eabalea accurately surprised, the confusion is related from the conflation of "root" (which folks assume to mean self-signed, self-issued, but is merely the apex of the chain) with "trust anchor" (the 'root' of trust, but not necessarily self-signed or self-issued).

**Comment by techliaison** [comment]
or the confusion can be from conflating self-signed and self-issued certificates, it clearly wanted a distinction to be made.
5280 states:
CA certificates may be further divided into three classes: cross-certificates, self-issued certificates, and self-signed certificates.

**Comment by sleevi** [comment]
While I'm wanting to avoid the game of "Stretch the RFC to support how you want to read it", reading a few sentences further states:
> Self-signed certificates are self-issued certificates where the digital signature may be verified by the public key bound into the certificate.

To ensure we don't distract too much from Wayne's original request:
Is the objection that Wayne's proposed language is not clear, or is the objection that it's undesirable, from a policy perspective, to require the disclosure of certificates issued by CAs in Mozilla's program?

**Comment by techliaison** [comment]
I am in favor of disclosing certificates appropriately and I feel that self-signed certificates should be disclosed for the purpose for which they were created - ie if created to be a new Root they should be disclosed as a potential new root.  I find a requirement to disclose them as an intermediate certificate very counter intuitive and misleading.

Disclosing the new key is done via the cross-certificate, assuming that is properly disclosed as an intermediate.  I just do not understand the benefit of requiring the disclosure of a certificate intended to be a new root, as an intermediate.  

It is still my understanding that when validating a cert path, the application should stop when it hits a certificate which was signed by private key associated with the public key signed and not try to keep on building a further path.
I know it can stop prior to hitting such a certificate if it encounters some other certificate considered to be a trust anchor.

I understand that any path can be evaluated if given as input, but I do not understand why one would want to disclose unintended paths.  Again the risk to the users of this "new CA" if you consider a CA a combination of Issuer DN & key, has already been disclosed via the disclosure of the actual intermediate cross-cert.

I am not objecting to disclosure of all CA certificates issued by CAs in Mozilla's program - I just think they should be disclosed for the correct purpose.

**Comment by sleevi** [comment]
There is no disclosure based on purpose, though. That is, Mozilla Policy
doesn't make the distinction on how the CA intends to use it - in fact, it
works very hard to require disclosure based on what it technically is. This
is to avoid issues with CAs being confused (as they have been in the past)
or gaming the system (as they have in the past) regarding intent. As a
consequence, all certificates that are CA certificates, and issued by a CA
directly or transitively included in Mozilla's program, must be
disclosed. The definition focuses on what the certificate _is_ - that is,
issued by - rather than how the holder of such a certificate intends to use
it.

The same disclosure requirements and policy requirements exist based on
capability, not intent. If it's capable of being used for X, it's subject
to the policies of X.

>

**Comment by techliaison** [comment]
and the CA has been disclosed via the intermediate cross-cert.

But I am willing to drop this discussion.

**Comment by sleevi** [comment]
Disclosure is by certificate, not by CA (that is, entity defined as
"Subject + SPKI").

If you imagine each CA{Subject, SPKI} tuple as a node within a graph, and
each edge identified by Certificate{Issuer, Signature}, then we want all
edges from all nodes in the graph disclosed, if they connect to another
node with CA=True. Edges that loop back to the same node are still Edges.

**Comment by benwilsonusa** [comment]
The third paragraph could be amended to read: "These requirements include all cross-certificates and self-signed certificates (e.g. "Issuer" DN is equivalent to "Subject" DN and public key is signed by the private key) that chain to a CA certificate that is included in Mozilla’s CA Certificate Program, and CAs must disclose such CA certificates in the CCADB."

**Comment by WilsonKathleen** [comment]
This tripped up another CA: https://bugzilla.mozilla.org/show_bug.cgi?id=1667844
